File: muscleyn/muscleyn/services/productService.ts
import api from "./api";

export const getProducts =
  async (

    page = 0,

    size = 10

  ) => {

    const response =

      await api.get(

        `/products?page=${page}&size=${size}`
      );

    return response.data;
};

export const getProductById =
  async (
    productId: number
  ) => {

    const response =

      await api.get(
        `/products/${productId}`
      );

    return response.data;
};

export const searchProducts =
  async (
    search?: string,
    categories?: string[],
    brands?: string[],
    goals?: string[],
    isBestSeller?: boolean,
    isOffer?: boolean,
    minPrice?: number,
    maxPrice?: number,
    page = 0,
    size = 10,
    sortBy = "id",
    direction = "desc"
  ) => {
    const params: any = {
      page,
      size,
      sortBy,
      direction,
    };
    if (search) params.search = search;
    if (categories && categories.length > 0) params.categories = categories.join(",");
    if (brands && brands.length > 0) params.brands = brands.join(",");
    if (goals && goals.length > 0) params.goals = goals.join(",");
    if (isBestSeller !== undefined) params.isBestSeller = isBestSeller;
    if (isOffer !== undefined) params.isOffer = isOffer;
    if (minPrice !== undefined) params.minPrice = minPrice;
    if (maxPrice !== undefined) params.maxPrice = maxPrice;

    const response = await api.get(
      "/products/search",
      { params }
    );

    return response.data;
};

export const getCategories =
  async () => {

    const response =

      await api.get(
        "/categories"
      );

    return response.data;
};

export const getBrands =
  async () => {

    const response =

      await api.get(
        "/brands"
      );

    return response.data;
};

export const getVariants =
  async (
    productId: number
  ) => {

    const response =

      await api.get(

        `/product-variants/product/${productId}`
      );

    return response.data;
};

File: muscleyn-backend/src/main/kotlin/com/muscleyn/backend/service/ProductService.kt
package com.muscleyn.backend.service

import com.muscleyn.backend.dto.CreateProductRequest
import com.muscleyn.backend.dto.ProductRequest
import com.muscleyn.backend.entity.Product
import com.muscleyn.backend.response.ProductResponse
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.web.multipart.MultipartFile

interface ProductService {

    fun createProduct(
        request: ProductRequest
    ): ProductResponse

    fun getAllProducts():
            List<ProductResponse>

    fun getAllProducts(
        page: Int,
        size: Int
    ): Page<ProductResponse>;

    fun searchProducts(
        search: String?,
        categories: List<String>?,
        brands: List<String>?,
        goals: List<String>?,
        isBestSeller: Boolean?,
        isOffer: Boolean?,
        minPrice: Double?,
        maxPrice: Double?,
        page: Int,
        size: Int,
        sortBy: String,
        direction: String
    ): Page<ProductResponse>

    fun getProductById(
        productId: Long
    ): ProductResponse

    fun updateProduct(

        productId: Long,

        request: ProductRequest

    ): ProductResponse

    fun deleteProduct(
        productId: Long
    )

    fun createProduct(

        request:
        CreateProductRequest,

        image:
        MultipartFile?,

        images: List<MultipartFile>?,

        productReport: MultipartFile?

    ): Product

    fun updateProduct(

        productId: Long,

        request: CreateProductRequest,

        image: MultipartFile?,

        images: List<MultipartFile>?,

        productReport: MultipartFile?

    ): Product

    fun deleteProductImage(imageId: Long)

    fun toggleBestSeller(productId: Long): ProductResponse

    fun toggleOffer(productId: Long): ProductResponse

    fun updateProductReport(
        productId: Long,
        productReport: MultipartFile,
        reportProteinPercentage: String?,
        reportHeavyMetal: String?,
        reportAminoAcidProfile: String?,
        reportMicrobialSafety: String?,
        reportTestDetails: String?
    ): ProductResponse

    fun deleteProductReport(productId: Long): ProductResponse
}